sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs

```diff
@@ -43,14 +47,64 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
 
         private static readonly ImmutableArray<KnownType> TrackedTypes = ImmutableArray.Create(KnownType.System_DirectoryServices_DirectoryEntry);
+
+        private static bool IsAllowedObject(ISymbol symbol, SyntaxNode node, SemanticModel semanticModel)
+        {
+            if (!symbol.GetSymbolType().Is(KnownType.System_DirectoryServices_AuthenticationTypes))
+            {
+                return false;
+            }
+
+            // In order to correctly detect all the cases we will need to refactor the rule and use symbolic execution.
+            // Until this is done, we can reduce the number of false positives by checking if the variable has one of the following
+            // values assigned in the parent scope: AuthenticationTypes.None, AuthenticationTypes.Anonymous, default or default(AuthenticationTypes).
+            var root = node.FirstAncestorOrSelf<MethodDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<ClassDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<StructDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<InterfaceDeclarationSyntax>() ??
+                       node.SyntaxTree.GetRoot();
+
+            return !HasUnsafeDeclaration(root, symbol, semanticModel) &&
+                   !HasUnsafeAssignment(root, symbol, semanticModel);
+        }
+
+        private static bool HasUnsafeAssignment(SyntaxNode root, ISymbol symbol, SemanticModel semanticModel) =>
+            root.DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => AssignmentLeftHasSameSymbol(assignment, symbol, semanticModel))
+                .Any(assignment => HasUnsafeConstantValue(assignment.Right, semanticModel));
+
+        private static bool AssignmentLeftHasSameSymbol(AssignmentExpressionSyntax assignment, ISymbol symbol, SemanticModel semanticModel) =>
+            assignment.Left is IdentifierNameSyntax identifierNameSyntax &&
+            identifierNameSyntax.NameIs(symbol.Name) &&
+            symbol.Equals(semanticModel.GetSymbolInfo(identifierNameSyntax).Symbol);
+
+        private static bool HasUnsafeDeclaration(SyntaxNode root, ISymbol symbol, SemanticModel semanticModel) =>
+            root.DescendantNodes()
+                .OfType<VariableDeclaratorSyntax>()
+                .Where(variableDeclarator => VariableDeclaratorHasSymbol(variableDeclarator, symbol, semanticModel))
+                .Any(variableDeclarator => HasUnsafeConstantValue(variableDeclarator.Initializer.Value, semanticModel));
+
+        private static bool VariableDeclaratorHasSymbol(VariableDeclaratorSyntax variableDeclarator, ISymbol symbol, SemanticModel semanticModel) =>
+            variableDeclarator.Identifier.ValueText == symbol.Name &&
+            symbol.Equals(semanticModel.GetDeclaredSymbol(variableDeclarator));
+
+        private static bool HasUnsafeConstantValue(SyntaxNode node, SemanticModel semanticModel) =>
```

Comment:
the logic is a bit hard to follow
- HasUnsafeConstantValue - negative in name
- `!IsValidAuthenticationType(authType)` - negation with positive name
- `IsValidAuthenticationType` contains negations inside `authType != AuthenticationTypesNone && authType != AuthenticationTypesAnonymous`

negative logic is harder to follow

if would be easier to read 
```
private static bool HasUnsafeConstantValue(SyntaxNode node, SemanticModel semanticModel) =>
 // ...
 // IsUnsafeAuthenticationType(authType);	 
```

and then the only negation would be in the names , and would be consistent